SS4 security changes and unit test fixes

fspringveldt opened the pull request:
This PR updates the module as per the recent SS4 security/member upgrades, and also does the following:
* Existing unit test fixes + introduces LDAPAuthenticator unit tests (more to follow later to improve the code coverage)
* Retires LDAPSecurityController and abdicates it's responsibility to Logout -, Login and ChangePassword handlers
* Session static ::get and ::set calls
* Updates QueuedJobs namespaces
* Linting fixes
* Refactors deprecated Member::currentUser and Member::logIn calls
* Refactors deprecated Session::set and -get calls
* SAML only changed at the surface for now. Will work more on this.

Commit:
FIX new security and fixed unit tests
* Updated all QueuedJobs references to Symbiote namespace
* Authentication is now an interface so updated all it's subclasses in this module
* Fixed broken unit tests

Commit:
FIX replaced static Session:: calls

Commit:
FIX removed typo from search location

Commit:
FIX lint fixes for lines > 120 chars

Commit:
FIX added composer install to travis build

Commit:
Switched to precise distro

In an effort to resolve travis issues

robbieaverill commented:
Travis failures logged here: https://github.com/silverstripe/silverstripe-travis-support/issues/48

Commit:
FIX Update travis config to be standalone

Commit:
FIX updated silverstripe-australia reference to symbiote

Commit:
FIX added cms modules mysite bootstrap to travis config

robbieaverill committed:
Add CodeSniffer to dev dependencies

robbieaverill committed:
Add codecov.io configuration

robbieaverill commented on src/Control/LDAPSecurityController.php:
```diff
@@ -58,7 +58,11 @@ public static function getPasswordResetLink($member, $autologinToken)
     public function ChangePasswordForm()
     {
-        return Object::create('SilverStripe\\ActiveDirectory\\Forms\\LDAPChangePasswordForm', $this, 'ChangePasswordForm');
+        return Object::create(
```
Object doesn't exist anymore.

robbieaverill reviewed (changes requested):
This pull request is not ready, it doesn't contain any logic for handling authentication. See Simon's docs here: https://github.com/silverstripe/silverstripe-framework/pull/7102

Commit:
Added LDAPAuthenticator unit tests

Commit:
Login and Logout handlers added for Authenticators

Commit:
DOC updated to reflect security updates

fspringveldt commented:
* Implemented the logic for handling the authentication no LDAPAuthenticator and SAMLAuthenticator. Introduced two LoginHandlers accordingly
* Added LDAPAuthenticator tests as this was easily mockable. SAML not so much.
@robbieaverill

robbieaverill commented on src/Authenticators/LDAPAuthenticator.php:
```diff
@@ -78,69 +80,70 @@ public static function get_login_form(Controller $controller)
-     * @throws HTTPResponse_Exception
+     * @param HTTPRequest $request
+     * @param ValidationResult|null $result
+     * @return bool|Member
```
It still returns void/null

robbieaverill commented on src/Authenticators/LDAPAuthenticator.php:
```diff
@@ -163,18 +166,39 @@ public static function authenticate($data, Form $form = null)
-     * @param null|Form $form
+     * @param HTTPRequest $request
      * @return null|Member
+     * @internal param null|Form $form
```
I'm not sure what the purpose of this is?

fspringveldt replied:
Yeah, that is weird. And I see it was introduced by me. Must've been an accidental paste. Will remove.

robbieaverill commented on src/Authenticators/LDAPAuthenticator.php:
```diff
@@ -163,18 +166,39 @@ public static function authenticate($data, Form $form = null)
+
+    /**
+     * @inheritdoc
+     */
```
You can just omit this if there's nothing to add, it will be inherited by default

robbieaverill commented on src/Authenticators/LDAPAuthenticator.php:
```diff
@@ -163,18 +166,39 @@ public static function authenticate($data, Form $form = null)
-            array_merge($data, ['Email' => $data['Login']]),
-            $form
-        );
+        return Authenticator::LOGIN | Authenticator::LOGOUT;
```
I'm reasonably sure we need to support changing passwords too, but I suppose that we could log that and address it separately

fspringveldt replied:
You're right!
I must check if the below is true and that bit too 👍 
```
SilverStripe\ActiveDirectory\Services\LDAPService:
  allow_password_change: true
```

robbieaverill replied:
I feel like this method should just have the bitmask in it, we can put the conditional logic into a more specific method for changing passwords, otherwise the way this module represents itself is not consistent depending on the user's settings. What do you think?

robbieaverill commented on src/Authenticators/LDAPLoginForm.php:
```diff
@@ -110,6 +111,17 @@ public function __construct($controller, $name, $fields = null, $actions = null,
+     */
+    public function getAuthenticatorName()
+    {
+        return _t(self::class . '.AUTHENTICATORNAME', 'LDAP');
```
`::class` won't work at the moment with the text collector, you should use `__CLASS__` instead

robbieaverill commented on src/Authenticators/LDAPMemberLoginHandler.php:
```diff
@@ -6,9 +6,10 @@
 use SilverStripe\Security\Member;
-use SilverStripe\Security\MemberLoginHandler;
+use SilverStripe\Security\MemberAuthenticator\LoginHandler;
+use SilverStripe\Security\MemberAuthenticator\MemberAuthenticator;
```
Was this a missed import? I don't see where it's used

fspringveldt replied:
Yeah unused. Will remove.

robbieaverill commented on src/Authenticators/SAMLAuthenticator.php:
```diff
@@ -5,11 +5,17 @@
 use SilverStripe\Security\Authenticator;
+use SilverStripe\Security\Member;
+use SilverStripe\Security\MemberAuthenticator\LoginHandler;
+use SilverStripe\Security\MemberAuthenticator\LDAPLogoutHandler;
```
This isn't used here

robbieaverill commented on src/Control/LDAPSecurityController.php:
```diff
@@ -70,7 +71,11 @@ public function lostpassword()
 
-        if (Config::inst()->get('SilverStripe\\ActiveDirectory\\Authenticators\\LDAPAuthenticator', 'allow_email_login') === 'yes') {
+        if (Config::inst()->get(
+            'SilverStripe\\ActiveDirectory\\Authenticators\\LDAPAuthenticator',
```
If you use `LDAPAuthenticator::class` you can keep this on one line

robbieaverill commented on src/Control/SAMLController.php:
```diff
@@ -154,12 +154,14 @@ public function metadata()
-        if (Session::get('BackURL') && Director::is_site_url(Session::get('BackURL'))) {
-            return $this->redirect(Session::get('BackURL'));
+        if ($this->getRequest()->getSession()->get('BackURL') &&
+            Director::is_site_url($this->getRequest()->getSession()->get('BackURL'))) {
```
If dropping a line I think it's nice if you take the `&&` operator with it, so it doesn't look like a statement on its own

fspringveldt replied:
I agree. Fixed this and the one a bit further down below 👍

robbieaverill commented on src/Forms/LDAPChangePasswordForm.php:
```diff
@@ -9,8 +9,9 @@
 use SilverStripe\Forms\TextField;
-use SilverStripe\Security\ChangePasswordForm;
 use SilverStripe\Security\Member;
+use SilverStripe\Security\MemberAuthenticator\ChangePasswordForm;
```
Where's this used?

fspringveldt replied:
The class extends it.

robbieaverill replied:
True that!

robbieaverill commented on tests/php/LDAPAuthenticatorTest.php:
```diff
@@ -0,0 +1,119 @@
+        $this->data = ['Login' => $member->Email, 'Email' => $member->Email, 'Password' => 'password'];
+        $result = $this->callAuthMethod();
+        $this->assertInstanceOf(Member::class, $result);
+        $this->assertEquals($result->Email, $member->Email);
```
I guess that `$member->Email` is the expectation, that should be the first argument

robbieaverill commented on tests/php/LDAPAuthenticatorTest.php:
```diff
@@ -0,0 +1,119 @@
+        $this->data = ['Login' => 'usernotfound', 'Password' => 'passwordnotfound'];
+        $this->callAuthMethod();
+        $this->assertFalse($this->result->isValid());
+        $this->assertContains($this->result->getMessages()[0]['message'], 'Username not found');
```
Switch arguments around

robbieaverill commented on .travis.yml:
```diff
@@ -1,28 +1,37 @@
+  - phpenv config-rm xdebug.ini
+
+  - composer install --prefer-dist
+  - composer require --prefer-dist --no-update silverstripe/framework:4.0.x-dev
```
We should use `silverstripe/recipe-core 1.0.x-dev` here which will centralise some of these dependencies

robbieaverill reviewed (changes requested):
Looks good overall! I've left a few comments/suggestion, will give it a test run on Monday

Commit:
Updated authenticate return type in docbloc

Commit:
Updated travis to require silverstripe/recipe-core 1.0.x-dev

Commit:
Removed rogue @internal phpdoc entry

Commit:
Added password change support to authenticator if settings et

Commit:
Doc update and LDAPLoginForm constructor args updated

Commit:
FIX updated ::class in _t() to __CLASS__ for text collector support

Commit:
FIX removed unused class imports

Commit:
FIX tidied up code and removed unused imports

Commit:
FIX refactored old Form::messageForForm() occurrences

Commit:
FIX Replaced class references

Commit:
FIX replaced deprecated Member::logIn()

Commit:
A bunch of PR comment fixes

Commit:
FIX LostPassword and ChangePassword handlers implemented.
* removed obsolete LDAPSecurityController and deferred it's methods to
 various handlers
* implemented LDAPLostPasswordHandler and LDAPChangePasswordHandler

Commit:
FIX added samaccountname to fake gateway

Commit:
FIX removed obsolute Member::currentUser() reference

Commit:
FIX username and email support for change password form

fspringveldt committed:
FIX fallbackAuthenticate doesn't have Email key set

fspringveldt committed:
FIX forgot password email template path

Commit:
FIX replaced obsolete LoggerInterface::warn calls

Commit:
FIX skip fallback auth if invalid credentials received upstream

Commit:
FIX using recipe-cms instead of core as module requires cms

Commit:
FIX refactored MemberGroup updates so it's optional upon login

fspringveldt commented:
@robbieaverill LDAP was successfully tested with the below configuration (adding credentials as needed to mysite/_config/ldap.yml):

```yaml
SilverStripe\ActiveDirectory\Authenticators\LDAPAuthenticator:
  fallback_authenticator: yes
  allow_email_login: yes
SilverStripe\Core\Injector\Injector:
    SilverStripe\Security\Security:
      properties:
          Authenticators:
            default: %$SilverStripe\ActiveDirectory\Authenticators\LDAPAuthenticator
SilverStripe\ActiveDirectory\Services\LDAPService:
  allow_password_change: true
SilverStripe\ActiveDirectory\Model\LDAPGateway:
  options:
    'host': ''
    'port': ''
    'username': ''
    'password': ""
    'accountDomainName': 'your.domain.com'
    'baseDn': 'DC=your,DC=domain,DC=com'
    'networkTimeout': 10
    'useSsl': 'TRUE'
    'accountCanonicalForm': 4
```

Commit:
FIX password send email shows recipient name and _t() updates

Commit:
FIX Ipmlemented password change handler

robbieaverill commented on src/Authenticators/LDAPChangePasswordHandler.php:
```diff
@@ -0,0 +1,163 @@
+            try {
+                $userData = $service->getUserByGUID($member->GUID);
+            } catch (Exception $e) {
+                Injector::inst()->get('Logger')->error($e->getMessage());
```
This changed a while back, it's now `Injector::inst()->get(LoggerInterface::class)` with `use Psr\Log\LoggerInterface`

fspringveldt replied:
Aah yes. Thanks.

Commit:
FIX removed needless dot if username and not email given

robbieaverill commented on src/Authenticators/LDAPChangePasswordHandler.php:
```diff
@@ -0,0 +1,163 @@
+                $member->FailedLoginCount = null;
+                $member->write();
+
+                if (!empty($_REQUEST['BackURL'])
```
It might be better to use `$this->getRequest()->requestVar('BackURL')` to avoid using the superglobal

robbieaverill commented on src/Authenticators/LDAPLostPasswordHandler.php:
```diff
@@ -0,0 +1,238 @@
+                ),
+                null,
+                null,
+                $this
```
I reckon we could consolidate this a little by only passing the first two arguments in each condition above, then below add `$loginField->setForm($this);`

robbieaverill commented on src/Control/SAMLController.php:
```diff
@@ -175,10 +181,20 @@ protected function getRedirect()
      */
     public function getLogger()
     {
         return Injector::inst()->get('Logger');
```
`LoggerInterface::class`

robbieaverill commented on src/Forms/LDAPChangePasswordForm.php:
```diff
@@ -19,37 +23,44 @@ class LDAPChangePasswordForm extends ChangePasswordForm
+     * @param \SilverStripe\Control\RequestHandler $controller
+     * @param string $name
+     * @param null $fields
+     * @param null $actions
```
Better to type hint what you're expecting rather than null. `FieldList`?

robbieaverill commented:
Probably not the fault of this pull request, but the LDAPMemberSyncTask is currently failing in some cases, e.g.:

```
ERROR [Warning]: in_array() expects parameter 2 to be array, null given
IN GET dev/tasks/LDAPMemberSyncTask
Line 686 in /Users/robbieaverill/dev/silverhub/activedirectory/src/Services/LDAPService.php
```

and

```
[2017-09-05 10:52:44] ManyManyList::add() doesn't accept unsaved records
```

Commit:
FIX LoggerInterface::class wasn't used to obtain logger

Commit:
FIX used object method instead of superglobal

Commit:
FIX added type hints and small refactor in lostPasswordForm method

Commit:
FIX skip mapped group removal if no groups are mapped

fspringveldt commented:
> Probably not the fault of this pull request, but the LDAPMemberSyncTask is currently failing in some cases, e.g.:

I've added a check to skip this step if the user doesn't have any mapped groups.

robbieaverill commented:
Looks pretty good. @halkyon @mateusz if one of you have some time would you mind doing a review of this?

mateusz commented on src/Services/LDAPService.php:
```diff
@@ -488,26 +494,29 @@ public function getLDAPGroupMembers($dn)
      * @param array|null $data If passed, this is pre-existing AD attribute data to update the Member with.
      *            If not given, the data will be looked up by the user's GUID.
+     * @param bool $updateGroups controls whether to run the resource-intensive group update function as well. This is
+     *                          skipped during login to reduce load.
```
I don't think it's a valid speed optimisation technique to remove a feature :-) This feature is needed - otherwise the users will have to wait until the sync task runs before they can log in, which is not acceptable.

Example scenario:
- user is added to LDAP directly by the system administrator
- sync task has not ran yet (say, because it's hourly, or never runs because the site relies on login-sync)
- user tries to log in
- authorisation fails - because group sync hasn't ran, and the user has no groups.

fspringveldt replied:
Thanks @mateusz. I had to add the switch locally as it was too slow during local testing. 
I've re-enabled it on LDAPAuthenticator::authenticate() on line 164 👍

mateusz commented:
Needs testing against LDAP and SAML.

Commit:
FIX re-enabled updateMemberFromLDAP upon authentication

Commit:
PSR2 linting fixes

robbieaverill commented:
I'm going to merge this ahead of module separation (see #104) at which point we will test and fix the LDAP components as necessary, and mark SAML as unstable for the time being.

patricknelson commented:
🎉
